### PR TITLE
Fix numerical stability of logsumexp

### PR DESCRIPTION
**Pull request by bengioe**
This is minor but may bite us in edge cases, the `logsoftmax` operation in `GraphActionCategorical` was essentially computing `log(exp(logits - b))` which is terrible for the numerical stability of the gradient. This PR is just replacing `log(exp(logits - b))` by `logits - b`.

Another edge case: this PR also adds default outputs for the max scatter reduction. This avoids the max value being 0 if there is nothing of a certain type and all other max values are less than zero 0.

This PR also adds tests that would have failed before it, and some repeated operations in `GraphActionCategorial` methods are now done with `_compute_batchwise_max`.

### CODE REVIEW

**Review by aebrer — APPROVED** [review]
looks good to me from my limited understanding
